src/DGtal/base/SetFunctions.h

```diff
@@ -0,0 +1,1469 @@
+      static inline const key_type& key( const value_type& value )
+      {
+        return value;
+      }
+    };
+
+    /// unordered map-like adapter.
+    template <typename Container>
+    struct ComparatorAdapter< Container, true, false, true > 
+    {
+      typedef typename Container::value_type      value_type;
+      typedef typename Container::key_type        key_type;
+      typedef KeyComparatorForPairKeyData
+        < std::less< key_type >, value_type >     LessThanPredicate;
+      typedef KeyComparatorForPairKeyData
+        < std::equal_to< key_type >, value_type > EqualPredicate;
+
+      static LessThanPredicate less( const Container& /* C */ )
+      {
+        return LessThanPredicate( std::less< key_type >() );
+      }
+      static EqualPredicate equal_to( const Container& /* C */ )
+      {
+        return EqualPredicate( std::equal_to< key_type >() );
+      }
+      static inline const key_type& key( const value_type& value )
+      {
+        return value.first;
+      }
+
+    };
+
+
+
+    /**
+     * Description of template class 'SetFunctions' <p> \brief Aim:
+     * Specialize set operations (union, intersection, difference,
+     * symmetric_difference) according to the given type of
+     * container. It uses standard algorithms when containers are
+     * ordered, otherwise it provides a default implementation.
+     *
+     * @tparam Container any type of container.
+     *
+     * @tparam associative tells if the container is associative
+     * (e.g. set, map, unordered_set, unordered_map).
+     *
+     * @tparam ordered tells if the container is ordered (e.g., set, map).
+     *
+     *
+     * Specialized implementations are marked with (S) in the list below.
+     *
+     * |--------------------|-------------|----------|----------|
+     * | Container          | associative | ordered  |  pair    |
+     * |--------------------|-------------|----------|----------|
+     * | vector             |   false     |  false   |  false   |
+     * | list               |   false     |  false   |  false   |
+     * | (to avoid)         |   false     |  false   |  true    |
+     * | sorted vector (S)  |   false     |  true    |  false   |
+     * | sorted list (S)    |   false     |  true    |  false   |
+     * | set (S)            |    true     |  true    |  false   |
+     * | map (S)            |    true     |  true    |  true    |
+     * | unordered_set (S)  |    true     |  false   |  false   |
+     * | unordered_map (S)  |    true     |  false   |  true    |
+     *
+     * @note For pair containers (like map and unordered_map), the
+     * data is not taken into account, which means that it can be lost
+     * in some (modifier) operations.
+     */
+    template <typename Container, bool associative, bool ordered>
+    struct SetFunctionsImpl
+    {
+      /** 
+       * Equality test. This version does not use the
+       * fact that the container is ordered.
+       * @param[in] S1 an input set.
```

Comment:
maybe you should state that container values are be sorted (with O(nlogn) cost for some containers) before comparing them.